[FeedManager] FeedManager balance report

ew3361zh opened the pull request:
<!-- Provide a short summary of the changes this pull request contains. -->

Establishes a reporting system for the feed manager to track feed in and feed out and checks it periodically throughout the simulation to ensure things are balanced.

### Context
<!-- Use this section to link this pull request to other issues, other pull requests, or mention people. -->
Issue(s) closed by this pull request: closes #2460

### What
<!-- Add more detail about the changes that are being made in the pull request. -->
Creates a system which tracks running totals during the simulation for:
- Amount of each feed requested by `Animal` module each day
- Amount of each purchased feed used to fulfill that request
- Amount of each farmgrown feed used to fulfill that request
- Amount of feed purchased on that day
- Amount of each farmgrown feed remaining in storage on that day
- Amount of each purchased feed remaining in storage on that day

These are added daily to running totals (specific FeedManager attributes outlined in `How` section) during the simulation which can be checked at any point.

Additionally it sets up daily reporting for:
- Amount of each farmgrown feed in storage
- Amount of each purchased feed in storage
 
These latter numbers being reported daily are important for EEE emissions calculations which will be addressed in the final part of issue #1782.

### Why
<!-- Discuss why the changes in this pull request are needed or important. -->
We need to be able to verify that the amount of feed being purchased, requested by the animal module, fed to animals, grown and stored on the farm, and currently in a storage is all in balance. This system sets up that reporting to `OutputManager`.

### How
<!-- Give an explanation of how this pull request addresses needed changes. -->
1. 4 new attributes are added to `FeedManager` to track the accounting of feed within the `FeedManager`:
   - ~`_feed_requests`~ `_cumulative_feed_requests` - a dict of feeds and amounts requested by the `Animal` module each day
   - ~`_purchased_feeds_fed`~ `_cumulative_purchased_feeds_fed` - a dict of amounts of purchased feeds used to fulfill the daily request from the `Animal` module
   - ~`_farmgrown_feeds_fed`~ `_cumulative_farmgrown_feeds_fed` - a dict of amounts of farmgrown feeds used to fulfill the daily request from the `Animal` module
   - ~`_purchased_feeds`~ `_cumulative_purchased_feeds` - a dict of the amounts of each type of feed purchased each day

2. Renamed the `FeedManager` `execute_daily_routine()` function to `report_feed_storage_levels()` to more aptly describe what it does which is to be a wrapper function for reporting both the current amounts of stored purchased feeds and stored farmgrown feeds.
3. Creates a function in `FeedManager` `report_feed_manager_balance()` that reports the totals of 4 new above attributes as well as the amounts of farmgrown and purchased feeds in storage on a particular day during the simulation. That function is called from `SimulationEngine` on the ration formulation schedule since it's not necessary to have a daily update on this balance report (but that would be available if someone wanted it).
4. Updates reporting to OM on feed to remove different other intermediary attempts at developing and implementing a feed manager balance report.
5. Renames the `FeedManager` `get_total_inventory()` function to -> `get_total_projected_inventory()` as it's more accurately a projection function, not getting the actual current total feed inventory. Also updates the names for the data it returns to more accurately reflect what that data is.

### Test plan
<!-- Explain how you tested the changes and how you know the code functions as expected. -->
I used this as my test filter and went through a few different feeds - you can run it without a specific feed if you want but I found it to be an overwhelming amount of information. `44` is, in the current default simulation, a good example of a feed that's both grown on the farm and purchased so you get examples of multiple aspects of the feed balance reporting in action.

```
{
    "multiple": [
        {
            "name": "Feed Balance",
            "filters": [
                "FeedManager.report_feed_manager_balance.*_44_.*",
                ".*_44.balance_storage_levels"
            ]
        }
    ]
}
```
Example output (I removed the caller class and function at the start of each of these to fit it on a single screenshot but the final two columns represent the amount of farmgrown 44 in storage and purchased 44 in storage remaining on those particular days. The distinction between farmgrown and purchased storage is clear if you don't remove the caller class and function from the variable names as I did here to try and fit it all on one screenshot):
<img width="1484" height="524" alt="Screenshot 2025-07-22 at 5 37 11 PM" src="https://github.com/user-attachments/assets/750a8c47-615b-491f-9934-5033edd0df8d" />

allisterakun commented on RUFAS/biophysical/feed_storage/feed_manager.py:
```diff
@@ -145,9 +145,26 @@ def translate_crop_config_name_to_rufas_id(
         return next_harvest_dates_rufas_ids
 
-    def _query_result_factory(self, crop_category: CropCategory, amount: float) -> QUERY_RESULT_DATA_TYPE:
+    def _query_result_factory(self, stored_feed: CropCategory | int, amount: float) -> QUERY_RESULT_DATA_TYPE:
```
Just curious, is this function really necessary?

ew3361zh replied:
I think this was a remnant of having multiple querying functions before and addressing some mypy errors. I can assess and probably remove as it seems excessive now.

allisterakun reviewed (approved):
Looks excellent, Niko! 
Just one comment about possible future improvements: 
It looks like we are reporting two sets of outputs daily: 
- the feed requested/available/purchased of the current day
- the total feed requested/purchased/fed up until the current day (along with the total stored farmgrown/purchased feed at the end of each day).

I'd suggest renaming these variables to be clearer about what values they represent, similar to how the manure module has daily vs accumulated variables. But this can be a separate issue.

matthew7838 commented on RUFAS/simulation_engine.py:
```diff
@@ -256,12 +261,12 @@ def _initialize_simulation(self) -> None:
         crop_config_to_rufas_ids_map = self.field_manager.get_crop_configs_to_rufas_ids()
 
         nutrient_standard = NutrientStandard(self.im.get_data("config.nutrient_standard"))
-        # Uncomment the below three lines for testing pup
-        # self.im.delete_data("config.nutrient_standard")
-        # print(self.im.get_data("config.nutrient_standard"))
-        # print(self.im.get_metadata("properties.config_properties.nutrient_standard"))
```
Thank you, I forgot to remove it before merging from the previous PR.

matthew7838 commented on RUFAS/biophysical/feed_storage/feed_manager.py:
```diff
@@ -121,11 +98,34 @@ def __init__(
+        self._feed_requests: dict[RUFAS_ID, float] = {feed.rufas_id: 0.0 for feed in self.available_feeds}
+        self._purchased_feeds_fed: dict[RUFAS_ID, float] = {feed.rufas_id: 0.0 for feed in self.available_feeds}
+        self._farmgrown_feeds_fed: dict[RUFAS_ID, float] = {feed.rufas_id: 0.0 for feed in self.available_feeds}
+        self._purchased_feeds: dict[RUFAS_ID, float] = {feed.rufas_id: 0.0 for feed in self.available_feeds}
```
Adding these to the attribute section could also clarify things a lot.

KFosterReed commented:
So far this is looking great! I can trace the connections between the feed purchases on the ration interval and daily time steps, the harvested feeds added to storage, the feeds requested/fed to animals, and the total feed available. I have a couple of thoughts/questions that are not necessarily things that need to be changed but could be things that I need to wrap my head around: 


- I don't understand the benefit of the running totals for the requested/purchased/fed. Is this something that I or another SME asked for? Is there a benefit from the software implementation side that I'm not seeing? For large farms and long simulations, these numbers will get very very large and, right now, I don't quite understand their purpose. 
- For the daily purchased amounts, in a regular simulation, they are all 0 which makes sense given the updates to the ration interval purchase amounts. What I am grappling with is that, looking at the daily outputs, on the days when there is a ration formulation interval purchase, the daily purchased feed is 0 when there actually was feed purchased that day, just not from the daily purchasing amount. 
- I think it would be useful to have a daily amount of feed removed/ used from both purchased and farm grown feed stores. This will be useful when we come back to adding shrink but also for enabling summations of the total feeds used over a period of time which, I think, will be a more flexible way to get to the information in the running totals I mentioned above.

ew3361zh commented:
>     * I don't understand the benefit of the running totals for the requested/purchased/fed. Is this something that I or another SME asked for? Is there a benefit from the software implementation side that I'm not seeing? For large farms and long simulations, these numbers will get very very large and, right now, I don't quite understand their purpose.

The running totals approach wasn't specifically requested, it was just my approach to be able to confirm the feed manager is in balance with the anaimal module and feed storage. Sending requested/purchased/farmgrown/fed numbers to the OM daily would be fairly taxing size wise.
We also get into the issue we saw leading up to ADSA that purchases are not made every day and farmgrown feeds are available on a seasonal schedule. Aligning the purchases/farmgrown amounts with the requests/amounts fed becomes challenging. 
So what these numbers do is allow us to do a periodic accounting check-in - similar as we do with processing nutrient degradations in feed storage (and on the same schedule). They are updated anytime one of the associated functions is called and they can be queried at any point in the simulation. This allows us to check if these numbers which updated on different schedules are in balance.

>     * For the daily purchased amounts, in a regular simulation, they are all 0 which makes sense given the updates to the ration interval purchase amounts. What I am grappling with is that, looking at the daily outputs, on the days when there is a ration formulation interval purchase, the daily purchased feed is 0 when there actually was feed purchased that day, just not from the daily purchasing amount.

Yea, what would be helpful here - I added `purchase type` to try and help differentiate where the purchase is originating from - would different naming be helpful or is it not helpful to differentiate between purchases types? I think the idea here is that because each purchase type is on such a different schedule, a daily feed purchase report would be difficult to put together and I'm not sure how informative it would be since it would majority be zeros. Right now you could sum the total amount of feed by purchase type. Let me know what data would be helpful to look at and I can think about how we might get at it.

>     * I think it would be useful to have a daily amount of feed removed/ used from both purchased and farm grown feed stores. This will be useful when we come back to adding shrink but also for enabling summations of the total feeds used over a period of time which, I think, will be a more flexible way to get to the information in the running totals I mentioned above.

~~I think we still have this on dev and I could re-institute it here.~~
Edit: this is currently still being reported to OM. In `FeedManager._deduct_feeds_from_inventory()` we have two local variables that track the amount of each feed by type that is removed from storage to fulfill the feed request:
```
total_purchased_feed_deductions: dict[RUFAS_ID, float] = {}
total_farmgrown_feed_deductions: dict[RUFAS_ID, float] = {}
```

These are reported to OM at the end of the function on a daily basis for each feed by type:
`f"farmgrown_feed_{feed_id}_total_amount_deducted"`
and
`f"purchased_feed_{feed_id}_total_amount_deducted"`

KFosterReed commented:
> ~I think we still have this on dev and I could re-institute it here.~ Edit: this is currently still being reported to OM. In `FeedManager._deduct_feeds_from_inventory()` we have two local variables that track the amount of each feed by type that is removed from storage to fulfill the feed request:
> 
> ```
> total_purchased_feed_deductions: dict[RUFAS_ID, float] = {}
> total_farmgrown_feed_deductions: dict[RUFAS_ID, float] = {}
> ```
> 
> These are reported to OM at the end of the function on a daily basis for each feed by type: `f"farmgrown_feed_{feed_id}_total_amount_deducted"` and `f"purchased_feed_{feed_id}_total_amount_deducted"`

So it looks like these outputs were actually deleted in this PR. I copied the lines from the `FeedManager._deduct_feeds_from_inventory()` function to report the outputs and was able to see them but am still having trouble reconciling the numbers. Looking at the corn grain (feed 44) in the example freestall scenario, here is what I am seeing:

- When there are no farmgrown feed deductions, the daily purchased feed deductions match the requested amount and the amount removed from the available amount. This is great! This part gets a passing grade from me :) 
- When farm grown feeds are deducted, however, things start to get a little murky. On the first day that both purchased feed and farmgrown feeds are deducted, the sum of the two matches the amount requested- this is what I would expect and makes sense. But on the following days, it looks like the amount of feed deducted from the purchased feed remains the same as the day before while the farm grown feed deduction is reported as matching the amount requested. This is rather minor and seems easily fixable
- The more important/confusing thing though is that I can't quite reconcile the reported amounts/timing of feeds deducted from the farm grown storage with the decreases in the reported amount of stored feeds. This seems to be a timing thing related to the first time that farm grown feeds are fed. Specifically, corn grain is reported to be stored for 33 days before anything is removed from that storage which is longer than the 30 day ration formulation interval and doesn't make sense to me. Then, the amount removed is equal to the sum of the first 26 days of farm grown feed deductions, which also doesn't make sense to me. After this first ration interval, the stored feed amounts are updated every 30 days (ration interval period) and the amount deducted matches the sum of the amounts reported to be deducted from the farm grown feeds.


Another more complicated challenge I am having with these feed storage outputs is that the method of storing and reporting feeds by harvest date very quickly gets difficult to parse and interpret, especially when it is not clear what simulation day is associated with each output. I don't think this is something we can/should fix in this PR but I would like to have a few exploratory conversations with SMEs that would be interested in using these outputs and think about how to improve/summarize this information in a more usable format.

ew3361zh commented:
> When farm grown feeds are deducted, however, things start to get a little murky. On the first day that both purchased feed and farmgrown feeds are deducted, the sum of the two matches the amount requested- this is what I would expect and makes sense. But on the following days, it looks like the amount of feed deducted from the purchased feed remains the same as the day before while the farm grown feed deduction is reported as matching the amount requested. This is rather minor and seems easily fixable

To clarify, what would be the expected removal strategy in this situation? Would you expect the amount requested to be split evenly between purchased feeds and farm-grown feeds if both are available? Or that purchased feeds would be prioritized over farm-grown feeds?

I thought I remembered discussing this being the order of removal:

1. Oldest-stored farm-grown feed
2. ...
3. Newest-stored farm-grown feed
4. Any purchased feed regardless of storage date because degradations aren't occurring with purchased feeds after storage.

What would be the ideal expected removal strategy?

I'm looking further into the second question and will get back to you!

KFosterReed commented:
> To clarify, what would be the expected removal strategy in this situation? Would you expect the amount requested to be split evenly between purchased feeds and farm-grown feeds if both are available? Or that purchased feeds would be prioritized over farm-grown feeds?
> 
> I thought I remembered discussing this being the order of removal:
> 
> 1. Oldest-stored farm-grown feed
> 2. ...
> 3. Newest-stored farm-grown feed
> 4. Any purchased feed regardless of storage date because degradations aren't occurring with purchased feeds after storage.
> 
> What would be the ideal expected removal strategy?


That is exactly it! Oldest farm grown --> newest farm grown --> purchased :)

ew3361zh commented:
> That is exactly it! Oldest farm grown --> newest farm grown --> purchased :)

So for this point...

> On the first day that both purchased feed and farmgrown feeds are deducted, the sum of the two matches the amount requested- this is what I would expect and makes sense. But on the following days, it looks like the amount of feed deducted from the purchased feed remains the same as the day before while the farm grown feed deduction is reported as matching the amount requested. This is rather minor and seems easily fixable

...is the issue that purchased feed is being deducted on that first day that farmgrown feed is also available when it should be 100% farmgrown feed as soon as it's available? Or is it a different issue?

ew3361zh commented:
>     * The more important/confusing thing though is that I can't quite reconcile the reported amounts/timing of feeds deducted from the farm grown storage with the decreases in the reported amount of stored feeds. This seems to be a timing thing related to the first time that farm grown feeds are fed. Specifically, corn grain is reported to be stored for 33 days before anything is removed from that storage which is longer than the 30 day ration formulation interval and doesn't make sense to me. Then, the amount removed is equal to the sum of the first 26 days of farm grown feed deductions, which also doesn't make sense to me. After this first ration interval, the stored feed amounts are updated every 30 days (ration interval period) and the amount deducted matches the sum of the amounts reported to be deducted from the farm grown feeds.

Can you share what report you are using to see this? I'm not sure I fully grasp it still but seeing some screenshots of what you're looking at would be helpful if you have them available.

> Another more complicated challenge I am having with these feed storage outputs is that the method of storing and reporting feeds by harvest date very quickly gets difficult to parse and interpret, especially when it is not clear what simulation day is associated with each output. I don't think this is something we can/should fix in this PR but I would like to have a few exploratory conversations with SMEs that would be interested in using these outputs and think about how to improve/summarize this information in a more usable format.

Yea this I definitely agree with. Maybe some renaming of variables would be helpful? Or we could be doing some calculations in the reporting function so the differences are spelled out more clearly.

I wrote out some notes about what the daily process currently is and the constraints inherent in how a simulation day unfolds:

I think part of the issue is when I set up reporting to OM in the simulation day process. I'll use an example day when it's a ration reformulation day since those are the days on which we're reporting feed manager balance. The order currently is:

1. Process degradations on farmgrown feeds in storage (leading to DM losses of those feeds)
2. Purchase feeds based on ration interval projections
   - total amount of feed purchased to date is updated
3. Report the balance of:
   - total amount of feed requested to date
   - total amount of feed purchased to date (includes above ration interval purchases)
   - total amount of farmgrown feed used to fulfill requests
   - total amount of purchased feed used to fulfill requests
   - total amount of purchased feed in storage
   - total amount of farmgrown feed in storage
4. Send that day's feed request to FeedManager.
5. FeedManager evaluates the feed request which involves checking current inventory (purchased and farmgrown combined) as well as daily purchase allowance.
   - total amount of feed requested to date is updated (this amount would not be part of the total amount of feed requested to date until the next ration interval so the amount reported earlier in the simulation day would exclude it).
6. If there are any daily purchases needed to supplement shortfalls from available inventory, those are made here. This almost certainly wouldn't happen on a ration reformulation day as the ration interval purchases would have happened earlier in the day so all feeds should be adequately supplied on this day.
   - total amount of purchased feed is updated (again, this amount would not be part of the total amount of feed purchased to date until the next ration interval)
7. FeedManager reduces the feeds from storage according `to the farm grown --> newest farm grown --> purchased` ordering.
   - total amount of farmgrown feed in storage is updated (not reflected in amount of farmgrown feed in storage reported earlier in day)  
   - total amount of purchased feed in storage is updated (not reflected in amount of purchased feed in storage reported earlier in day)
   - total amount of farmgrown feed fed is updated (not reflected in either the total farmgrown feed used to fulfill requests reported earlier in the day)
   - total amount of purchased feed fed is updated (same issue as for farmgrown feed)
   
So because the reporting of the FeedManager balance happens on the ration reformulation schedule, we don't have that same day's reductions included nor any daily purchases if they apply.

KFosterReed commented:
> Can you share what report you are using to see this? I'm not sure I fully grasp it still but seeing some screenshots of what you're looking at would be helpful if you have them available.

Here is the filter I was using: 

`{
    "name": "Stored Corn Grain Total",
    "type": "plot",
    "filters": [
      "Dry.*record_storage_totals.*fresh_mass.*",
      "Dry.*record_storage_totals.*dry_matter_mass.*"
    ],
    "expand_data": true,
    "use_fill_value_in_gaps": false,
    "use_fill_value_at_end": false
},
 {
            "name": "Feed Requests and Purchases",
            "filters": [
                "FeedManager.purchase_feed.*44_amount_purchased.*",
                "FeedManager.manage_daily_feed_request.*44_requested_amount.*",
                "FeedManager.manage_daily_feed_request.*44_available_amount.*",
                "FeedManager._deduct_feeds_from_inventory.purchased_feed_44_total_amount_deducted",
                "FeedManager._deduct_feeds_from_inventory.farmgrown_feed_44_total_amount_deducted",
                "FeedManager.report_daily_purchases.*44_amount_purchased.*"
            ],
    "expand_data": true,
    "use_fill_value_in_gaps": false,
    "use_fill_value_at_end": false
        }`

And here is a link to the [output in a google sheet](https://docs.google.com/spreadsheets/d/1ds22zWyYDv-4tputQaMOCbrrakdU8ijfi6KMrQcXZZ0/edit?usp=sharing)

for the `available_amount` in this filter, do you know if this includes both farm grown and purchased feeds? And when this value would be updated in the steps you described above?

ew3361zh commented:
> for the available_amount in this filter, do you know if this includes both farm grown and purchased feeds? And when this value would be updated in the steps you described above?

This DOES include all available farm grown and purchased feeds for that day combined into one total for both sources by feed id (i.e. so for feed 44 - corn grain - the amount available would be purchased corn feed in storage + farmgrown corn feed in storage).
And it gets logged to OM during step 5 - so within `FeedManager`.`manage_daily_feed_request()`. 
The only days this will not be updated are:
1. If the request for a particular feed is not fulfillable by either what's available in storage (purchased and farmgrown) or by the daily purchase allowance.
2. If a particular feed was not part of the feed request for that day (I don't know if this ever happens tbh but the possibility is there if the feed request varied over the course of the simulation to not include a certain feed on a certain day).

KFosterReed commented:
I think I am finally able to reconcile all the additions and deductions! Thank you for sharing that order of operations @ew3361zh! Very helpful :) 

The way all this info is reported is still too confusing for my liking but at least it is (mostly) accurate. I have two requests before approving this PR and then we can think more about how to make the farm grown feed reports more clear. 

1. For the stored corn grain amounts (`"Dry.*record_storage_totals.*fresh_mass.*"`) it doesn't appear like the data expansion is working. This makes it challenging to reconcile timelines so can this be fixed to include that data expansion functionality?
2. For the reported purchased feed deduction amounts, it looks like it is replicating the last amount deducted (see my google sheet), even when the full feed request is also being deducted from the farm grown feed. If this can be changed to be 0 when everything is pulled from storage that would be ideal. 

After that, I think this PR is good to go!

ew3361zh commented:
@KFosterReed I'm fixing some broken tests this morning but I believe I have addressed your feedback and done a little more to try and clarify things for either now or the future. Here are the updates:

1. Added the `fresh_mass` to the reporting of farmgrown feeds in storage to OM. Previously it was just dry matter mass.
2. It will now report 0 when the amount of a particular feed deducted on a particular is 0 - previously it would not report anything and would default to data padding based on the last available deduction amount.
3. Enforced the deduction strategy we discussed earlier - (oldest farm grown --> newest farm grown --> purchased).
4. Moved the reporting of the daily storage amounts from the start of the day to after ration interval purchasing has been done which makes the reporting make a little more sense (avoids lag of a purchase happening on a particular day and it not being reported as being in storage on that same day).

I also took some time yesterday to try to get the different datap oints available from OM organized in a way that made more sense to me. I don't know if this is necessarily the PR to fully dive into this as I think there are a number of other things waiting on this PR to be merged but hopefully helpful to see and start evaluating the naming of the variables, and the necessity and necessary frequency of all these data points for what SMEs need to know about feed manager balance.

Here's that report:
[docs.google.com/spreadsheets/d/1r-vDlw2B5dE_J3tZXK7Ylok695d5KHNS5cOW9O2q-OQ/edit?gid=1812856468#gid=1812856468](https://docs.google.com/spreadsheets/d/1r-vDlw2B5dE_J3tZXK7Ylok695d5KHNS5cOW9O2q-OQ/edit?usp=sharing)

Here's my updated report filter I used to get this data:

```
{
    "name": "Balance",
    "filters": [
    "FeedManager.manage_daily_feed_request.*44_requested_amount.*",
    "FeedManager.manage_daily_feed_request.*44_available_amount.*",
    ".*_44.*daily_storage_levels",
    "FeedManager._deduct_feeds_from_inventory.purchased_feed_44_total_amount_deducted",
    "FeedManager._deduct_feeds_from_inventory.farmgrown_feed_44_total_amount_deducted",
    "FeedManager.report_daily_purchases.*44_amount_purchased.*",
    "FeedManager.purchase_feed.*44_amount_purchased.*",
    "FeedManager.*purchased_feed_44_fed.*",
    "FeedManager.*farmgrown_feed_44_fed.*"
],
    "expand_data": true,
    "use_fill_value_in_gaps": false,
    "use_fill_value_at_end": false
}
```
